File: 0904-fruit-into-baskets/0904-fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        int l=0, ans=0;
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int r=0; r<fruits.length; ++r){
            map.put(fruits[r], map.getOrDefault(fruits[r], 0)+1);
            
            while(map.size()>2){
                map.put(fruits[l], map.get(fruits[l])-1);
                if(map.get(fruits[l])==0){
                    map.remove(fruits[l]);
                }
                l+=1;
            }
            ans = Math.max(ans, r-l+1);
            
        }
        return ans;
    }
}

File: 0904-fruit-into-baskets/0904-fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        left, count, ans = 0, collections.defaultdict(int), 0
        for right, val in enumerate(fruits):
            count[fruits[right]]+=1
            
            while len(count)>2:
                count[fruits[left]]-=1
                if not count[fruits[left]]:
                    count.pop(fruits[left])
                left+=1
            ans = max(ans, right-left+1)
        return ans



# class Solution:
#     def totalFruit(self, fruits: List[int]) -> int:
#         i, j, ans, mp = 0, 0, 0, {}
        
#         while j<len(fruits):
#             if fruits[j] in mp: mp[fruits[j]]+=1
#             else: mp[fruits[j]]=1
            
#             if len(mp)<=2:
#                 j+=1
#                 ans=max(ans, j-i)
#             else:
#                 mp[fruits[i]]-=1
#                 if mp[fruits[i]]==0: mp.pop(fruits[i])
#                 i+=1
#                 mp[fruits[j]]-=1
#                 if mp[fruits[j]]==0: mp.pop(fruits[j])
                
#         return ans
        


# class Solution:
#     def totalFruit(self, fruits: List[int]) -> int:
#         i, j, ans, mp = 0, 0, 0, {}
        
#         while j<len(fruits):
#             if fruits[j] in mp: mp[fruits[j]]+=1
#             else: mp[fruits[j]]=1
#             j+=1
            
#             if len(mp)<=2: ans=max(ans, j-i)
#             else:
#                 while len(mp)>2 and i<=j:
#                     mp[fruits[i]]-=1
#                     if mp[fruits[i]]==0: mp.pop(fruits[i])
#                     i+=1
        
#         return ans
        